src/safeds/data/tabular/typing/_column_type.py

```diff
@@ -31,17 +47,42 @@ def _from_numpy_data_type(data_type: np.dtype) -> ColumnType:
         NotImplementedError
             If the given data type is not supported.
         """
-        if data_type.kind in ("u", "i"):
-            return Integer()
-        if data_type.kind == "b":
-            return Boolean()
-        if data_type.kind == "f":
-            return RealNumber()
-        if data_type.kind in ("S", "U", "O", "M", "m"):
-            return String()
 
-        message = f"Unsupported numpy data type '{data_type}'."
-        raise NotImplementedError(message)
+        def column_type_of_type(cell_type: Any) -> ColumnType:
+            if cell_type == int or cell_type == np.int64 or cell_type == np.int32:
+                return Integer(is_nullable)
+            if cell_type == float or cell_type == np.float64 or cell_type == np.float32:
+                return RealNumber(is_nullable)
+            if cell_type == bool:
+                return Boolean(is_nullable)
+            if cell_type == str:
+                return String(is_nullable)
+            if cell_type is NoneType:
+                return Nothing()
+            else:
+                message = f"Unsupported numpy data type '{cell_type}'."
+                raise NotImplementedError(message)
+
+        result: ColumnType = Nothing()
+        is_nullable = False
+        for cell in data:
+            if result == Nothing():
+                result = column_type_of_type(type(cell))
+                if type(cell) is NoneType:
+                    is_nullable = True
+                    result._is_nullable = is_nullable
+            if result != column_type_of_type(type(cell)):
+                if type(cell) is NoneType:
+                    is_nullable = True
+                    result._is_nullable = is_nullable
+                elif (isinstance(result, Integer) and isinstance(column_type_of_type(type(cell)), RealNumber)) or (
+                    isinstance(result, RealNumber) and isinstance(column_type_of_type(type(cell)), Integer)
+                ):
+                    result = RealNumber(is_nullable)
+                else:
+                    result = Anything(is_nullable)
+
```

Comment:
```suggestion
            if isinstance(cell, float) and np.isnan(cell):
                is_nullable = True
                result._is_nullable = is_nullable

```

Reply:
See https://github.com/Safe-DS/Stdlib/pull/432#pullrequestreview-1525308460 for explanation of this suggestion